### PR TITLE
[Feature] Support read postion delete file for iceberg v2

### PR DESCRIPTION
**Pull request by caneGuy**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

implementation for iceberg v2 table reading of parquet-format
related with: [https://github.com/StarRocks/starrocks/issues/3452]

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by caneGuy** [comment]
run starrocks_admit_test

File: be/src/formats/parquet/file_reader.cpp [hunk 1]
```diff
@@ -33,6 +33,10 @@
 
 namespace starrocks::parquet {
 
+FileReader::FileReader(int chunk_size, RandomAccessFile* file, uint64_t file_size,
+                       std::set<std::int64_t> need_skip_rowids)
```
Comment (dirtysalt) [inline be/src/formats/parquet/file_reader.cpp:37]: pass const reference.

File: be/src/formats/parquet/group_reader.cpp [hunk 2]
```diff
@@ -30,6 +30,13 @@ namespace starrocks::parquet {
 constexpr static const LogicalType kDictCodePrimitiveType = TYPE_INT;
 constexpr static const LogicalType kDictCodeFieldType = TYPE_INT;
 
+GroupReader::GroupReader(GroupReaderParam& param, int row_group_number, std::set<std::int64_t> need_skip_rowids,
+                         int64_t row_group_first_row)
```
Comment (dirtysalt) [inline be/src/formats/parquet/group_reader.cpp:34]: pass const reference.  too many move is not good idea.

File: be/src/formats/parquet/group_reader.h [hunk 3]
```diff
@@ -137,6 +139,9 @@ class GroupReader {
 
     // row group meta
     std::shared_ptr<tparquet::RowGroup> _row_group_metadata;
+    std::int64_t _row_group_first_row = 0;
+    std::set<std::int64_t> _need_skip_rowids;
```
Comment (dirtysalt) [inline be/src/formats/parquet/group_reader.h:143]: yes, since this  set is immutable, use const ref is good.

**Comment by dirtysalt** [comment]
Please add some test cases.

**Review by Smith-Cruise — COMMENTED** [review]
You add some ut for it

File: be/src/exec/iceberg/iceberg_delete_builder.cpp [hunk 4]
```diff
@@ -34,6 +35,26 @@ static const IcebergColumnMeta k_delete_file_path{
...
+                                                  &(IcebergDeleteFileMeta::get_delete_file_pos_slot())};
+    auto iter = new IcebergDeleteFileIterator();
+    RETURN_IF_ERROR(iter->init(_fs, timezone, delete_file_path, file_length, slot_descriptors, true));
+    std::shared_ptr<::arrow::RecordBatch> batch;
+    while (iter->has_next()) {
+        batch = iter->next();
+        ::arrow::StringArray* file_path_array = static_cast<arrow::StringArray*>(batch->column(0).get());
+        ::arrow::Int64Array* pos_array = static_cast<arrow::Int64Array*>(batch->column(1).get());
+        for (size_t row = 0; row < batch->num_rows(); row++) {
+            if (file_path_array->Value(row) != _datafile_path) {
```
Comment (dirtysalt) [inline be/src/exec/iceberg/iceberg_delete_builder.cpp:50]: `!=` or `==` ?

Comment (dirtysalt) [reply be/src/exec/iceberg/iceberg_delete_builder.cpp:50]: _datafile_path is fullpath or relative path? Is that string very long?

Comment (caneGuy) [reply be/src/exec/iceberg/iceberg_delete_builder.cpp:50]: it is Full URI of a data file with FS scheme

**Comment by caneGuy** [comment]
run BE UT

File: be/src/exec/hdfs_scanner_parquet.cpp [hunk 5]
```diff
@@ -14,14 +14,25 @@
...
 #include "util/runtime_profile.h"
 
 namespace starrocks {
 
 static const std::string kParquetProfileSectionPrefix = "Parquet";
 
 Status HdfsParquetScanner::do_init(RuntimeState* runtime_state, const HdfsScannerParams& scanner_params) {
+    if (!scanner_params.deletes.empty()) {
+        auto iceberg_delete_builder =
+                new IcebergDeleteBuilder(scanner_params.fs, scanner_params.path, scanner_params.conjunct_ctxs,
```
Comment (zombee0) [inline be/src/exec/hdfs_scanner_parquet.cpp:30]: Hi, I can't find where it is released, maybe smart pointer is better.

File: be/src/exec/iceberg/iceberg_delete_builder.cpp [hunk 6]
```diff
@@ -34,6 +35,26 @@ static const IcebergColumnMeta k_delete_file_path{
 static const IcebergColumnMeta k_delete_file_pos{
         .id = INT32_MAX - 102, .col_name = "pos", .type = TPrimitiveType::BIGINT};
 
+Status ParquetPositionDeleteBuilder::build(const std::string& timezone, const std::string& delete_file_path,
+                                           int64_t file_length, std::set<int64_t>* need_skip_rowids) {
+    std::vector<SlotDescriptor*> slot_descriptors{&(IcebergDeleteFileMeta::get_delete_file_path_slot()),
+                                                  &(IcebergDeleteFileMeta::get_delete_file_pos_slot())};
+    auto iter = new IcebergDeleteFileIterator();
```
Comment (zombee0) [inline be/src/exec/iceberg/iceberg_delete_builder.cpp:42]: ditto

File: be/test/exec/iceberg/iceberg_delete_builder_test.cpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+protected:
+
+    std::string _parquet_delete_path = "./be/test/exec/test_data/parquet_scanner/parquet_delete_file.parquet";
+    std::string _parquet_data_path = "parquet_data_file.parquet";
+
+    std::set<std::int64_t> _need_skip_rowids;
+};
+
+TEST_F(IcebergDeleteBuilderTest, TestParquetBuilder) {
+    auto parquet_builder = new ParquetPositionDeleteBuilder(FileSystem::Default(), _parquet_data_path);
```
Comment (zombee0) [inline be/test/exec/iceberg/iceberg_delete_builder_test.cpp:57]: ditto

File: be/src/formats/parquet/file_reader.cpp [hunk 8]
```diff
@@ -459,7 +464,11 @@ Status FileReader::_init_group_readers() {
                 continue;
             }
 
-            auto row_group_reader = std::make_shared<GroupReader>(_group_reader_param, i);
+            if (i > 0) {
+                row_group_first_row += _file_metadata->t_metadata().row_groups[i].num_rows;
+            }
```
Comment (zombee0) [inline be/src/formats/parquet/file_reader.cpp:469]: count offset even not selected

**Review by zombee0 — COMMENTED** [review]
Left LGTM

**Review by zombee0 — APPROVED** [review]

File: be/src/formats/parquet/group_reader.h [hunk 9]
```diff
@@ -139,6 +140,9 @@ class GroupReader {
 
     // row group meta
     std::shared_ptr<tparquet::RowGroup> _row_group_metadata;
+    std::int64_t _row_group_first_row = 0;
```
Comment (Smith-Cruise) [inline be/src/formats/parquet/group_reader.h:143]: `_row_group_first_row_pos` better?

File: be/src/formats/parquet/group_reader.cpp [hunk 10]
```diff
@@ -79,6 +82,18 @@ Status GroupReader::get_next(ChunkPtr* chunk, size_t* row_count) {
     Filter chunk_filter(count, 1);
     DCHECK_EQ(active_chunk->num_rows(), count);
 
+    // row id filter
+    if ((nullptr != _need_skip_rowids) && !_need_skip_rowids->empty()) {
+        std::int64_t current_chunk_base_row = _row_group_first_row + _raw_rows_read - count;
```
Comment (Smith-Cruise) [inline be/src/formats/parquet/group_reader.cpp:87]: just use `int64_t`

File: be/src/formats/parquet/group_reader.cpp [hunk 11]
```diff
@@ -79,6 +82,18 @@ Status GroupReader::get_next(ChunkPtr* chunk, size_t* row_count) {
...
+        auto start_iter = lower_bound(_need_skip_rowids->begin(), _need_skip_rowids->end(), current_chunk_base_row);
+        auto end_iter =
+                upper_bound(_need_skip_rowids->begin(), _need_skip_rowids->end(), current_chunk_base_row + count);
+        for (; start_iter != end_iter; start_iter++) {
+            chunk_filter[*start_iter] = 0;
```
Comment (Smith-Cruise) [inline be/src/formats/parquet/group_reader.cpp:92]: I think this code is wrong, chunk_filter size is based on chunk size, its maximum size is 4096, but `*start_iter`'s value is pos in the whole parquet file, you can't do it directly.

**Review by Smith-Cruise — COMMENTED** [review]
Can you add ut for multi row group and multi-page in each row group?

**Review by Smith-Cruise — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Pull request merged**